vermouth/tests/test_name_moltype.py

```diff
@@ -0,0 +1,151 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Test for the NameMolType processor.
+"""
+
+from collections import defaultdict
+import itertools
+from glob import glob
+import os.path
+import subprocess
+import pytest
+from hypothesis import given, assume
+from hypothesis import strategies as st
+from vermouth import System
+from vermouth.processors.name_moltype import NameMolType
+from .molecule_strategies import random_molecule
+from .datafiles import PDB_HB
+
+@st.composite
+def molecules_and_moltypes(draw, max_moltypes=4, min_size=0, max_size=None,
+                           molecule_kwargs={}):
+    """
+    Generates a list of molecules and the list of the expected moltypes.
+
+    Parameters
+    ----------
+    draw:
+        Internal for hypothesis.
+    max_moltypes: int
+        Maximum number of different moltypes.
+    min_size: int
+        Minimum number of molecules.
+    max_size: int, optional
+        Maximum number of molecules. If `None`, behaves the same way as
+        :func:`hypothesis.strategies.lists`.
+    molecule_kwargs: dict
+        Arguments to pass to :func:`random_molecule`.
+
+    Returns
+    -------
+    hypothesis.searchstrategy.lazy.LazyStrategy
+        A list of molecules and a list of moltype names corresponding to these
+        molecules as they are expected to be assigned by
+        :class:`vermouth.processors.name_moltype.NameMolType`.
+    """
+    if max_size is not None and max_size < min_size:
+        raise ValueError('max_size ({}) must be greater than min_size ({}), or None.'
+                         .fornat(max_size, min_size))
+
+    # We first generate the list of expected moltypes, we can then use it as
+    # keys to draw the molecules with shared strategies. The moltypes must
+    # match what NameMolType would generate, so they must have the form
+    # molecule_0, molecule_1, molecule_x, where the index is increasing based
+    # on the first occurence of the molecule type.
+    # pre_moltypes will be the moltype names before we rename them based on
+    # their order.
+    n_moltypes = draw(st.integers(min_value=1, max_value=max_moltypes))
+    unique_pre_moltypes = range(n_moltypes)
+    pre_moltypes_sampled = draw(st.lists(st.sampled_from(unique_pre_moltypes),
+                                         min_size=min_size, max_size=max_size))
+    if not pre_moltypes_sampled:
+        return [], []
+
+    moltype_index = 0
+    pre_to_moltype = {pre_moltypes_sampled[0]: 'molecule_{}'.format(moltype_index)}
+    moltypes = []
+    for pre_moltype in pre_moltypes_sampled:
+        if pre_moltype not in pre_to_moltype:
+            moltype_index += 1
+            pre_to_moltype[pre_moltype] = 'molecule_{}'.format(moltype_index)
+        moltypes.append(pre_to_moltype[pre_moltype])
```

Comment:
I have the initial feeling this strategy is finished here. But I'll read on

Reply:
Without testing the assumption, the result of the strategy cannot be trusted. Without the copy, it is unusable.